site/content/blog/2021-04-16-migrating-operator-custom-resource.md

```diff
@@ -0,0 +1,148 @@
+```
+
+This way we can perform the migration in the following way:
+- When the controller for `ClusterInstallation` sees that the `spec.migrate` is set to `true`, it stops regular reconciliation of the CR and starts the migration by converting the old resource to the new one. 
+- Immediately after the `Mattermost` resource is created, the controller for `Mattermost` sees it and starts to adjust existing resources like Services and Deployments to the new CR, making small changes and overriding owner references.
+- When we are sure that the controller for `Mattermost` successfully finished its work and Mattermost Pods are ready to serve traffic, the old `ClusterInstallation` is deleted, and voila!
+- As in the second phase the controller is just checking new Pods health, it can just do it once and requeue the resource for latter reconciliation while starting the migration for the next one, which gives us ability to migrate multiple resources at a time.
+
+The migration process has another advantage, that is if anything goes wrong it can be easily reverted just by setting `spec.migrate` back to `false` and removing the newly created `Mattermost` CR. The controller for `ClusterInstallation` will then claim the resources back and continue to monitor them.
+
+
+### Dealing with immutable fields
+
+"Exchanging" the Kubernetes resources ownership between two Custom Resources worked fine in most cases but not for all of them.
+
+Old Mattermost Deployments created by the Operator used the following `spec.selector`:
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+...
+spec:
+  selector:
+    matchLabels:
+      app: mattermost
+      v1alpha1.mattermost.com/installation: my-mattermost
+      v1alpha1.mattermost.com/resource: my-mattermost
+...
+```
+
+It is understandable that when we set out to migrate to version `v1beta` we wanted to get rid of `v1alpha1` from everywhere and change it as well.
+
+Kubernetes Deployments have some awesome features like [rolling updates](https://kubernetes.io/docs/tutorials/kubernetes-basics/update/update-intro/) that we use all the time when we change environment variables, versions, or other configuration of our installations. It allows for updating Pods sequentially keeping some of them running while others are being updated.
+
+Unfortunately, the `spec.selector` field is immutable so we could not just update the Deployment. 
+We also cannot have two Deployments of the same name in the same namespace and we did not want to change the names of resources created by the Operator.
+Simply running the [Client Go](https://github.com/kubernetes/client-go) equivalent of `kubectl delete deployment...` and creating it from scratch was not an option either as it would cause the deletion of all Pods running Mattermost and result in a brief downtime to Mattermost instance managed by the Mattermost Operator and it would violate one of our objectives.
```

Comment:
```suggestion
Simply running the [Client Go](https://github.com/kubernetes/client-go) equivalent of `kubectl delete deployment...` and creating it from scratch wasn't an option either as it would cause the deletion of all Pods running Mattermost and result in a brief downtime to Mattermost instance managed by the Mattermost Operator and it would violate one of our objectives.
```